### PR TITLE
instantiate2

### PR DESCRIPTION
**Pull request by taitruong**

### CODE REVIEW

File: packages/ics721/src/execute.rs [hunk 1]
```diff
@@ -258,8 +259,24 @@ where
         let instantiate = if CLASS_ID_TO_NFT_CONTRACT.has(deps.storage, class.id.clone()) {
             vec![]
         } else {
+            let class_id = ClassId::new(class.id.clone());
```
Comment (taitruong) [inline packages/ics721/src/execute.rs:262]: part 1 of instantiate2: using class_id as salt,

File: packages/ics721/src/execute.rs [hunk 2]
```diff
@@ -299,13 +317,24 @@ where
...
+        let mut instantiate_msg = cw721_base::msg::InstantiateMsg {
             name: class.id.clone().into(),
             symbol: class.id.clone().into(),
             minter: env.contract.address.to_string(),
-        })
+        };
+
+        // unwrapped to collection data and in case of success, set name and symbol
+        if let Some(binary) = class.data.clone() {
+            let class_data_result: StdResult<CollectionData> = from_json(binary);
```
Comment (taitruong) [inline packages/ics721/src/execute.rs:329]: part 2, instantiate2: using symbol and name, in case class data holds collection data

File: packages/ics721/src/ibc.rs [hunk 3]
```diff
@@ -253,17 +252,8 @@ where
...
-                let cw721::ContractInfoResponse { name, .. } = deps
-                    .querier
-                    .query_wasm_smart(cw721_addr.clone(), &cw721::Cw721QueryMsg::ContractInfo {})?;
-                let class_id = ClassId::new(name);
-
-                // Save classId <-> contract mappings.
-                CLASS_ID_TO_NFT_CONTRACT.save(deps.storage, class_id.clone(), &cw721_addr)?;
-                NFT_CONTRACT_TO_CLASS_ID.save(deps.storage, cw721_addr.clone(), &class_id)?;
+                // cw721 addr has already been stored, here just check whether it exists, otherwise a NotFound error is thrown
+                let class_id = NFT_CONTRACT_TO_CLASS_ID.load(deps.storage, cw721_addr.clone())?;
```
Comment (taitruong) [inline packages/ics721/src/ibc.rs:256]: part 3, no need storing class id in name and symbol, instead: once instantiated by sub msg, query, whether return addr fits with addr created before using instantiate2.

File: packages/ics721/src/testing/integration_tests.rs [hunk 4]
```diff
@@ -90,68 +95,205 @@ fn no_init(
 ) {
 }
 
-#[derive(Debug)]
-struct Bech32AddressGenerator {
-    pub hrp: String,
+#[derive(Default)]
+pub struct MockAddressGenerator;
+
+impl AddressGenerator for MockAddressGenerator {
```
Comment (taitruong) [inline packages/ics721/src/testing/integration_tests.rs:101]: biggest changes are in test, using MockAddressGenerator and MockApiBech32. In generator, predictable_contract_address() function covers instantiate2.

Also all tests had to be refactored and cleanup for better readability.

File: contracts/sg-ics721/src/execute.rs [hunk 5]
```diff
@@ -36,45 +36,54 @@ impl Ics721Execute for SgIcs721Contract {
...
+                    // ics721 creator is used, in case of none
+                    {
+                        let ics721_contract_info = deps
+                            .querier
+                            .query_wasm_contract_info(env.contract.address.to_string())?;
+                        instantiate_msg.collection_info.creator = ics721_contract_info.creator;
+                    }
+                }
+                // set name and symbol
+                instantiate_msg.symbol = class_data.symbol;
```
Comment (humanalgorithm) [inline contracts/sg-ics721/src/execute.rs:82]: nice looks like you can take name and symbol right from the class data object?

File: contracts/sg-ics721/src/testing/integration_tests.rs [hunk 6]
```diff
@@ -91,68 +95,205 @@ fn no_init(
...
+        let parts: Vec<&str> = salt_to_string
+            .trim_matches(|c| c == '[' || c == ']')
+            .split(',')
+            .collect();
+        // Convert each part to a u8 and collect them into a Vec<u8>
+        let salt: Vec<u8> = parts
+            .iter()
+            .map(|part| part.trim().parse().unwrap())
+            .collect();
+        let canonical_addr = instantiate2_address(checksum, creator, &salt)?;
```
Comment (humanalgorithm) [inline contracts/sg-ics721/src/testing/integration_tests.rs:135]: the magic is here

File: contracts/sg-ics721/src/testing/integration_tests.rs [hunk 7]
```diff
@@ -91,68 +95,205 @@ fn no_init(
...
+impl MockApiBech32 {
+    pub fn new(prefix: &'static str) -> Self {
+        Self { prefix }
     }
 }
 
-#[cw_serde]
-pub struct CustomClassData {
-    pub foo: Option<String>,
+impl Api for MockApiBech32 {
```
Comment (humanalgorithm) [inline contracts/sg-ics721/src/testing/integration_tests.rs:168]: Maybe could move all of these setup classes into a different file

File: packages/ics721/src/execute.rs [hunk 8]
```diff
@@ -258,8 +259,24 @@ where
...
+                .map_err(|_| StdError::generic_err("Could not calculate addr"))?;
+            let cw721_addr = deps.api.addr_humanize(&canonical_cw721_addr)?;
+
+            // Save classId <-> contract mappings.
+            CLASS_ID_TO_NFT_CONTRACT.save(deps.storage, class_id.clone(), &cw721_addr)?;
+            NFT_CONTRACT_TO_CLASS_ID.save(deps.storage, cw721_addr.clone(), &class_id)?;
+
             let message = SubMsg::<T>::reply_on_success(
-                WasmMsg::Instantiate {
+                WasmMsg::Instantiate2 {
```
Comment (humanalgorithm) [inline packages/ics721/src/execute.rs:279]: This is the main message that we are trying to get back in this PR right?

File: packages/ics721/src/execute.rs (hunk 2 above)
Comment (humanalgorithm) [reply packages/ics721/src/execute.rs:329]: Who is calling this init_msg function and what Class object are they passing to it?

File: contracts/sg-ics721/src/testing/integration_tests.rs (hunk 7 above)
Comment (taitruong) [reply contracts/sg-ics721/src/testing/integration_tests.rs:158]: No, it is not as easy as it looks like, because this code is in sg-ics721 contract and also in ics721 package. So there will be duplicates in any case. Besides it is just test code, so here it is fine - in all other cases I wouldnt like duplicates neither :).

Comment (humanalgorithm) [reply contracts/sg-ics721/src/testing/integration_tests.rs:158]: Okay yeah, since its test code

File: packages/ics721/src/execute.rs [hunk 9]
```diff
@@ -258,8 +260,28 @@ where
...
+            let mut hasher = Sha256::new();
+            hasher.update(class_id.as_bytes());
+            let salt = hasher.finalize().to_vec();
+            // Get the canonical address of the contract creator
+            let canonical_creator = deps.api.addr_canonicalize(env.contract.address.as_str())?;
+            // get the checksum of the contract we're going to instantiate
+            let CodeInfoResponse { checksum, .. } = deps
+                .querier
+                .query_wasm_code_info(CW721_CODE_ID.load(deps.storage)?)?;
+            let canonical_cw721_addr = instantiate2_address(&checksum, &canonical_creator, &salt)?;
```
Comment (humanalgorithm) [inline packages/ics721/src/execute.rs:276]: Can get a quick explanation of what is going on here?

File: packages/ics721/src/query.rs [hunk 10]
```diff
@@ -81,20 +81,17 @@ pub trait Ics721Query {
...
-        )? else {
-        // Token metadata is set unconditionaly on mint. If we have no
-        // metadata entry, we have no entry for this token at all.
-        return Ok(None)
+        let Some(token_metadata) =
+            TOKEN_METADATA.may_load(deps.storage, (class_id.clone(), token_id.clone()))?
+        else {
+            // Token metadata is set unconditionaly on mint. If we have no
+            // metadata entry, we have no entry for this token at all.
+            return Ok(None);
```
Comment (humanalgorithm) [inline packages/ics721/src/query.rs:89]: If there is no token metadata we just return a None. Does the caller see that as an error? What does Ok(None) mean to the contract that is calling this?

File: ts-relayer-tests/ci-scripts/osmosis/template/.osmosisd/config/app.toml [hunk 11]
```diff
@@ -10,6 +10,10 @@
 # specified in this config (e.g. 0.25token1;0.0001token2).
 minimum-gas-prices = "0uosmo"
 
+# The maximum gas a query coming over rest/grpc may consume.
```
Comment (humanalgorithm) [inline ts-relayer-tests/ci-scripts/osmosis/template/.osmosisd/config/app.toml:13]: Why we have an Osmosis script in ics721?

File: packages/ics721/src/execute.rs (hunk 9 above)
Comment (taitruong) [reply packages/ics721/src/execute.rs:276]: This is mainly explained in docs here: https://github.com/CosmWasm/cosmwasm/blob/main/packages/std/src/addresses.rs#L270-L309

Essentially it uses salt for creating a unique address. Since `class_id` is unique for ics721, it is an ideal candidate for salt. NFT module has a exactly same approach: instantiated nft contract is a hash of class id.

File: packages/ics721/src/query.rs (hunk 10 above)
Comment (taitruong) [reply packages/ics721/src/query.rs:89]: ics721 has prepared for handling token metadata. prepared means it is stored in ics721 rn - since core `cw721` package does not support it per se. see zeke's comment in `callback_mint()` in execute.rs:

```rs
    fn callback_mint(
        &self,
        deps: DepsMut,
        class_id: ClassId,
        tokens: Vec<Token>,
        receiver: String,
    ) -> Result<Response<T>, ContractError> {
        let receiver = deps.api.addr_validate(&receiver)?;
        let cw721_addr = CLASS_ID_TO_NFT_CONTRACT.load(deps.storage, class_id.clone())?;

        let mint = tokens
            .into_iter()
            .map(|Token { id, uri, data }| {
                // We save token metadata here as, ideally, once cw721
                // supports on-chain metadata, this is where we will set
                // that value on the debt-voucher token. Note that this is
                // set for every token, regardless of if data is None.
                TOKEN_METADATA.save(deps.storage, (class_id.clone(), id.clone()), &data)?;
...
```

Flow is like this:
- on send from source, metadata is loaded (which is currently None)
- pass to target as part of `NonFungibleTokenPacketData`
- save metadata in ics721 on target chain

File: contracts/sg-ics721/src/execute.rs (hunk 5 above)
Comment (taitruong) [reply contracts/sg-ics721/src/execute.rs:82]: Some background here: ics721 need to store `class id <> nft contract` map entries (e.g. for determining incoming NFTs are back- (=burn) or forward transfer (=mint).

Before usage of `instantiate2` we **had** to store class id in collection, since address is unknown yet. and on instantiation reply we have address but no class id, so had to store it as part of collection.
Which led to the side effect frontends not having source name and symbol.

By using `instantiate2` ics721 now knows address - even before instantiation. So it can store name and symbol provided by source chain.

File: contracts/sg-ics721/src/testing/integration_tests.rs (hunk 6 above)
Comment (taitruong) [reply contracts/sg-ics721/src/testing/integration_tests.rs]: Correct, this is executed when `WasmMsg::Instantiate2` is called.

File: packages/ics721/src/query.rs (hunk 10 above)
Comment (humanalgorithm) [reply packages/ics721/src/query.rs:89]: I just want to be clear here, so ics721. is storing onchain metadata on ics721? As in,  when someone moves their nft from A to B, metadata is saved on chain B's ics721 contract. In our original discussion of this issue, we talked about querying original chain to get the onchain metadata from its original home chain. Does this mean we have abandoned that method?

File: ts-relayer-tests/ci-scripts/osmosis/template/.osmosisd/config/app.toml (hunk 11 above)
Comment (humanalgorithm) [reply ts-relayer-tests/ci-scripts/osmosis/template/.osmosisd/config/app.toml:13]: Nice, fyi @jhernandezb

File: packages/ics721/src/execute.rs (hunk 8 above)
Comment (taitruong) [reply packages/ics721/src/execute.rs:287]: Yes, ser

File: packages/ics721/src/execute.rs [hunk 12]
```diff
@@ -164,6 +164,9 @@ where
             return Err(ContractError::Unauthorized {});
         }
 
+        // cw721 doesn't support on-chain metadata yet
+        // here NFT is transferred to another chain, NFT itself may have been transferred to his chain before
+        // in this case ICS721 may have metadata stored
         let token_metadata = TOKEN_METADATA
             .may_load(deps.storage, (class.id.clone(), token_id.clone()))?
```
Comment (taitruong) [inline packages/ics721/src/execute.rs:171]: during `receive_nft()` (starting point from nft contract to ics721 for transfer) read  TOKEN_METADATA (in case it has been provided on previous, initial transfer to this chain), this is used for transferring to next chain. Not part of this PR. just outlining what happens here.

File: packages/ics721/src/execute.rs [hunk 13]
```diff
@@ -385,10 +388,10 @@ where
...
             .into_iter()
             .map(|Token { id, uri, data }| {
-                // We save token metadata here as, ideally, once cw721
-                // supports on-chain metadata, this is where we will set
-                // that value on the debt-voucher token. Note that this is
-                // set for every token, regardless of if data is None.
+                // Source chain may have provided token metadata, so we save token metadata here
+                // Note, once cw721 doesn't support on-chain metadata yet - but this is where we will set
+                // that value on the debt-voucher token once it is supported.
+                // Also note that this is set for every token, regardless of if data is None.
```
Comment (taitruong) [inline packages/ics721/src/execute.rs:394]: This  TOKEN_METADATA part is on receival on target chain, here we store metadata, for later re-transfer. This is not used on mint, since cw721 doesnt support on-chain metadata (yet).

File: packages/ics721/src/query.rs (hunk 10 above)
Comment (taitruong) [reply packages/ics721/src/query.rs:89]: No, we didn't abandoned it. What we store here is on initial transfer and storing data provided from `NonFungibleTokenPacketData`. Also added another commit with some more comments on metadata. Pls check also my other 2 comments on TOKEN_METADATA. Please note, this is not part of this PR - only added some comments in source for clarifications.

File: packages/ics721/src/execute.rs (hunk 1 above)
Comment (shanev) [reply packages/ics721/src/execute.rs:266]: Smart!

**Review by shanev — APPROVED** [review]

**Review by jhernandezb — APPROVED** [review]

**Review by humanalgorithm — APPROVED** [review]
